cg.go

```diff
@@ -0,0 +1,293 @@
+	Beta(grad, gradPrev, dirPrev []float64) float64
+}
+
+// CG implements the nonlinear conjugate gradient method for solving nonlinear
+// unconstrained optimization problems [1]. It is a line search method that
+// generates the search directions d_k according to the formula
+// d_{k+1} = -∇f_{k+1} + β_k*d_k,   d_0 = -∇f_0.
+// Different variants of the conjugate gradient method differ in the choice
+// of the parameter β_k. Conjugate gradient method requires no matrix storage
+// and is faster than the gradient descent method, but L-BFGS is usually more
+// efficient and robust.
+//
+// CG implements a restart strategy that takes the steepest descent direction
+// (i.e., d_{k+1} = -∇f_{k+1}) whenever any of the following conditions holds.
+// * A certain number of iterations has elapsed without a restart. This number
+//   depends on the dimension of the problem and is controllable via
+//   IterationRestartFactor.
+// * The angle between the gradients at two consecutive iterations ∇f_k and
+//   ∇f_{k+1} does not satisfy a certain criterion.
+// * β_k given by Variant.Beta() is equal to zero.
+//
+// The line search for CG must yield step sizes that satisfy the strong Wolfe
+// conditions at every iteration, otherwise the generated search direction
+// might fail to be a descent direction. The line search should be more
+// accurate compared to the case of Newton-like methods, which can be achieved
+// by setting the gradient constant in the strong Wolfe conditions to a small
+// value.
+//
+// [1] William Hager, Hongchao Zhang, A survey of nonlinear conjugate gradient
+//     methods. Pacific Journal of Optimization, 2 (2006), pp. 35-58.
+type CG struct {
+	// LinesearchMethod must satisfy the strong Wolfe conditions at every
+	// iteration. If LinesearchMethod == nil, an appropriate default is chosen
+	// with the gradient constant set to 0.1.
+	LinesearchMethod LinesearchMethod
+	// Variant implements the particular CG formula for computing β_k.
+	// If Variant is nil, an appropriate default is chosen.
+	Variant CGVariant
+	// InitialStep estimates the initial line search step size, because the CG
+	// method does not generate well-scaled search directions.
+	// If InitialStep is nil, an appropriate default is chosen.
+	InitialStep StepSizer
+
+	// The iteration is restarted by taking the steepest descent direction
+	// whenever ceil(IterationRestartFactor*(dimension of the problem))
+	// iterations have elapsed without a restart.
+	// For medium and large-scale problems IterationRestartFactor should be set
+	// to 1 and greater values should be used for low-dimensional problems.
+	// Care must be taken because if the resulting number is 1, the iteration
+	// will be restarted at every iteration and the CG method will reduce to
+	// the steepest descent method.
+	// If IterationRestartFactor is 0, it will be set to 6.
+	IterationRestartFactor float64
+	// The method is restarted whenever the gradients at two consecutive
```

Comment:
AngleRestartThreshold sets the threshold angle for restart. If the angle between two consecutive gradients is greater than AngleRestartThreshold
  formula
the decent direction is the same as the gradient. A value of AngleRestartThreshold closer to 1 will reduce the number of restarts.